main/src/ca/uwaterloo/flix/language/phase/Uncurrier.scala

```diff
@@ -0,0 +1,243 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.Expression._
+import ca.uwaterloo.flix.language.ast.SimplifiedAst.{LoadExpression, StoreExpression}
+import ca.uwaterloo.flix.language.{CompilationError, GenSym}
+import ca.uwaterloo.flix.language.ast.{Type, _}
+import ca.uwaterloo.flix.util.{InternalCompilerException, Validation}
+import ca.uwaterloo.flix.util.Validation._
+
+/**
+  * A phase that merges curried definitions into definitions which take multiple
+  * arguments.
+  *
+  * For example, consider
+  *
+  *   |def add(x: Int): Int -> Int = {
+  *   |   z -> x + z
+  *   |}
+  *   |
+  *   |def a: Int = (add(3))(4)
+  *
+  * The uncurrier will create a function
+  *
+  *   |def add$uncurried(x: Int, z: Int): Int = {
+  *   |    x + z
+  *   |}
+  *
+  * And change the call site to
+  *   |def a: Int = add$uncurried(3,4)
+  */
+object Uncurrier extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+
+  def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
+    implicit val _ = flix.genSym
+    val startTime = System.nanoTime()
+
+    val defns = root.definitions.foldLeft(Map.empty[Symbol.DefnSym, SimplifiedAst.Definition.Constant])(Definitions.mkUncurriedDefs)
+    val uncurriedDef = defns map {case (k,v) => k -> Definitions.uncurry(v, root) }
+
+    val currentTime = System.nanoTime()
+    val time = root.time.copy(uncurrier = currentTime - startTime)
+    root.copy(definitions = uncurriedDef).copy(time = time).toSuccess
+  }
+
+  object Definitions {
+    /**
+      * Create uncurried versions of all curried functions
+      *
+      * @param map The old map of function definitions
+      * @param pair The entry which to make uncurried versions of
+      * @return A new map with the uncurried versions added
+      */
+    def mkUncurriedDefs(map: Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant], pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      map ++ mkUncurriedDef(pair)
+    }
+
+    /**
+      * Creates all the uncurried versions of a function
+      *
+      * @param pair The dfnsym definition pair to base the functions on
+      * @return A map containing all the new definitions
+      */
+    def mkUncurriedDef(pair: (Symbol.DefnSym, SimplifiedAst.Definition.Constant))(implicit genSym: GenSym): Map[Symbol.DefnSym, SimplifiedAst.Definition.Constant] = {
+      val sym = pair._1
+      val curriedDef = pair._2
+      curriedDef.exp match {
```

Comment:
This does not seem recursive to me. Does this create uncurried versions?

E.g. x -> y -> z -> w  should probably have the versions:

(x, y) -> z -> w
(x, y, z) -> w